HostIr Integration #4

nsarka opened the pull request:
This PR fixes some bugs with HostIr integration into FusionExecutorCache. The list of fixes are:

1) Expressions within the SegmentedGroup were out of order. They get sorted during gpu lowering. So, I topologically sorted them after adding them to the HostIrContainer

2) CompileFusionParallel launches compileKernel with multiple threads from a pool. However, the main thread does not wait on completion before continuing. I added fix so that it will wait.

3) During the preseg pass stage, Sets are added between expressions with a different sharding. These sets needed lowering into a communication. I added the missing lowering.

With these changes, many (most?) of the MultiDevice tests and also DistributedTransformer tests are passing. The failing tests seem to be because of an issue with loop splitting. Here is an example:

```
TEST_F(MultiDeviceTest, nick_hostir_5) {
  Fusion fusion, fusion_cloned;
  FusionGuard fg(&fusion);
  TensorView* x = makeContigTensor(4);
  x->outer_split(0, 2);
  fusion.addInput(x);
  std::cout << "x transforms: " << std::endl;
  x->printTransforms();
  fg.setCurFusion(&fusion_cloned);
  IrCloner ir_cloner(&fusion_cloned);
  auto in_clone = ir_cloner.clone(fusion.inputs());
  std::cout << "cloned_x transforms: " << std::endl;
  in_clone[0]->as<TensorView>()->printTransforms();
/*
x transforms:
 logical domain : (iS0{i0}, iS1{i2}, iS2{i3}, iS3{i4})
 contiguity: t t t t
  Outer split: iS0{i0} by factor 2 -> iS4{2}, iS5{( ceilDiv(i0, 2) )}          <----- outer split
 loop domain : (iS4{2}, iS5{( ceilDiv(i0, 2) )}, iS1{i2}, iS2{i3}, iS3{i4})
cloned_x transforms:
 logical domain : (iS0{i0}, iS1{i2}, iS2{i3}, iS3{i4})
 contiguity: t t t t                                                                                       <----- no outer split, cloner does not propagate it
 loop domain : (iS4{2}, iS5{i6}, iS1{i2}, iS2{i3}, iS3{i4})
*/
}
```

After cloning exprs into the hostir container, splits are not propagated.

wujingyue commented on csrc/host_ir/container.cpp:
```diff
@@ -54,6 +55,11 @@ void HostIrContainer::setKernelExecutor(
 
+void HostIrContainer::sortExprs()
+{
+  this->top_level_exprs_ = groupExprPrintSorting(this->top_level_exprs_);
```
Did you try https://github.com/NVIDIA/Fuser/blob/4d2db45698435920a90036c4bf1912a94808b6eb/csrc/iter_visitor.cpp#L904 or https://github.com/NVIDIA/Fuser/blob/4d2db45698435920a90036c4bf1912a94808b6eb/csrc/iter_visitor.cpp#L892? groupExprPrintSorting hasn't been optimized for speed. It's fine to use it for debugging but I'd be careful with using it elsewhere

wujingyue commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -445,10 +449,39 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
-          hic->pushBackTopLevelExprs(cloned_expr);
+        const bool is_resharding = std::any_of(
+            group_to_run->exprs().begin(), group_to_run->exprs().end(), [](auto expr) {
+              return isResharding(expr);
```
isResharding is expensive (it uses simplifyExpr eg), so I'd try to avoid it whenever possible. I believe you can check group_to_run->schedulerType() instead.

wujingyue commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -445,10 +449,39 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
+          HostIrLower lower;
+          for (auto* expr :
+              lower.lower(ir_cloner.clone(group_to_run->exprs().at(0)))) {
+            // Allocate the recv buffers of communications
```
Can this change go to a different PR? https://google.github.io/eng-practices/review/developer/small-cls.html

nsarka replied:
This part is copied from https://github.com/NVIDIA/Fuser/blob/f58617e8edfa13ad5e040046f3549d94402b3a45/csrc/host_ir/lower.cpp#L646 actually. Is it worth another PR?

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -25,15 +25,17 @@ using testing::Property;
-          &ExecutorAbstract::isA<HostIrExecutor>,
-          IsTrue()))))
-      << "failed to compile to a HostIrContainer with Communications";
+  if (!isOptionEnabled(EnableOption::HostIrLowering)) {
```
Sure -- this is a valid change. But can you come up with a check for host IR lowering? For example, check the top level instructions in a HostIrContainer has communications?

wujingyue commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -415,6 +418,7 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
           thread_pool_error_message = ss.str();
         }
       });
+      getThreadPool()->waitWorkComplete();
```
Great! Can you make this a separate PR and add @rdspring1 to review?

wujingyue commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -458,6 +494,8 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
       hic->addOutput(ir_cloner.clone(out));
     }
+
+    hic->sortExprs();
```
FYI, segments themselves are toposorted already (cf. group_run_order). So it's also faster to sort each segment than sort the whole thing. `mlogm + nlogn < (m+n)log(m+n)`

nsarka commented:
PR opened for waiting for compilation threads: https://github.com/NVIDIA/Fuser/pull/4139

nsarka commented:
> It's great to hear many tests are passing behind the scene. Is it possible to add some tests to this PR? For example, it could be some existing tests but with host IR lowering also enabled (cf. TEST_P) or could be some new tests to test_host_ir_integration to catch potential regression?

Updated with TEST_P on tests in `tests/cpp/test_multidevice_lower_communication.cpp`

nsarka commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -427,6 +447,17 @@ void FusionKernelRuntime::compileFusionParallel(KernelArgumentHolder args) {
     num_live_args_after_segment_runs_.emplace_back((int64_t)args.size());
   }
 
+  if (num_groups != 1 && !isOptionDisabled(DisableOption::ParallelCompile)) {
```
This is just moved from below

nsarka commented on csrc/runtime/fusion_kernel_runtime.h:
```diff
@@ -143,6 +143,10 @@ class FusionKernelRuntime {
 
   const std::vector<std::unique_ptr<ExecutorAbstract>>& executors() const;
 
+  const std::unique_ptr<hir::HostIrEvaluator>& getHostIrEvaluator() const {
```
This is for 
```
  if (isOptionEnabled(EnableOption::HostIrLowering)) {
    EXPECT_EQ(runtime->getHostIrEvaluator()->canRun(), "");
  } else {
    ...
  }
```

nsarka commented:
@wujingyue I removed the host ir lowering code and replaced it with a TODO. I also updated the tests to just pass false for enable hostir lowering. The next PR i can re-enable them, and re-add the hostir lowering code.

nsarka commented:
!test

wujingyue commented on csrc/runtime/fusion_kernel_runtime.cpp:
```diff
@@ -196,6 +199,28 @@ flatbuffers::Offset<serde::FusionKernelRuntime> FusionKernelRuntime::serialize(
+  {
+    auto
+        [/*IrCloner*/ group_ir_cloner,
+         /*std::unique_ptr<Fusion>*/ group_fusion] = fusion->makeFusion(group);
```
makeFusion is unfortunately more expensive than you might think. It copies the **complete** fusion rather than just Exprs in the group, which is unnecessary but has been there for legacy reasons. 

Did you try StmtSort::getExprsBetween or similar helpers?

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -26,14 +26,18 @@ namespace {
-      runtime->executors(),
-      Each(Pointer(Property(
-          "is a HostIrExecutor",
-          &ExecutorAbstract::isA<HostIrExecutor>,
```
This was intended to verify the fusion is compiled to communications. Recall that in the non host IR code path, HostIrExecutor is for and only for communications. 

Therefore, a better check would be to verify the number of Communication Exprs in the HostIrEvaluator's top-level expression list. In gmock, you can do this with `Contains(e).Times(n)`. See https://github.com/NVIDIA/Fuser/blob/6ff60e2a320733a2f49de57007d6bb45000107cd/tests/cpp/test_move_split_cat.cpp#L545 for an example.

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -56,11 +60,19 @@ using InOutMesh = std::pair<DeviceMesh, DeviceMesh>;
-                        public testing::WithParamInterface<InOutMesh> {};
+class LowerGatherTest
+    : public MultiDeviceTest,
+      public testing::WithParamInterface<std::tuple<InOutMesh, bool>> {};
```
```suggestion
      public testing::WithParamInterface<std::tuple<DeviceMesh, DeviceMesh, bool>> {};
```
I don't see a good reason for nesting. This will probably simplify the structured bindings below.

nsarka replied:
I think this is here because we want certain pairs of DeviceMesh, and not every pair that Combine(InMesh, OutMesh) would give

wujingyue reviewed (commented):
LGTM with comments

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -91,19 +109,47 @@ TEST_P(LowerGatherTest, ) {
+        testing::Values(false)), // TODO: testing::Bool() after implementing
+                                 // communication lowering
+    [](const testing::TestParamInfo<std::tuple<InOutMesh, bool>>& info) {
+      const auto& meshes = std::get<0>(info.param);
+      const auto& in_mesh = meshes.first;
+      const auto& out_mesh = meshes.second;
+      const auto enable_hir = std::get<1>(info.param);
```
Can you do something similar to line 75?

nsarka replied:
I would, but the compiler interprets the ] bracket in the tuple unpack as the closing of the lambda expression. I fiddled with it a bit, but didn't find a way to make it work.

wujingyue replied:
<html><head></head><body>
<h2>❌ Why structured bindings don't work in <code inline="">NameGenerator</code></h2>
<pre><code class="language-cpp">INSTANTIATE_TEST_SUITE_P(
    MyTest,
    MyParamTest,
    ::testing::Values(std::make_tuple(42, true)),
    [](const ::testing::TestParamInfo&lt;ParamType&gt;&amp; info) {
        auto [x, y] = info.param;  // ❌ ERROR: structured bindings not allowed
        return "SomeName";
    }
);
</code></pre>
<h3>❓ Why?</h3>
<ul>
<li>
<p>Structured bindings (<code inline="">auto [x, y] = ...</code>) are <strong>only allowed at block scope</strong> (i.e., inside a function body).</p>
</li>
<li>
<p>The <strong>lambda parameter of <code inline="">INSTANTIATE_TEST_SUITE_P</code> is used in a context where structured bindings are not legal</strong>, due to how macros and template instantiations are expanded.</p>
</li>
<li>
<p>You're passing a <strong>templated lambda</strong> into a macro, and the compiler can’t always deduce or allow destructuring that way.</p>
</li>
</ul>
<hr>
</body></html>

Instead, you can create a separate method that uses structured bindings and let the name generator call that. It probably dedups because there are several test suites in this file whose names are generated in the same way. 

```
diff --git a/tests/cpp/test_multidevice_lower_communication.cpp b/tests/cpp/test_multidevice_lower_communication.cpp
index 3587a534..f19975d5 100644
--- a/tests/cpp/test_multidevice_lower_communication.cpp
+++ b/tests/cpp/test_multidevice_lower_communication.cpp
@@ -179,6 +179,26 @@ TEST_P(LowerScatterTest, ) {
   }
 }

+namespace {
+std::string nameFromTuple(const std::tuple<InOutMesh, bool>& param) {
+  auto&& [meshes, enable_hir] = param;
+  auto&& [in_mesh, out_mesh] = meshes;
+
+  std::stringstream ss;
+  ss << "InMesh";
+  for (auto id : in_mesh.vector()) {
+    ss << "_" << id;
+  }
+  ss << "_OutMesh";
+  for (auto id : out_mesh.vector()) {
+    ss << "_" << id;
+  }
+  ss << (enable_hir ? "_HirEnabled" : "_HirDisabled");
+
+  return ss.str();
+}
+} // namespace
+
 INSTANTIATE_TEST_SUITE_P(
     HostIrLowering,
     LowerScatterTest,
@@ -188,23 +208,7 @@ INSTANTIATE_TEST_SUITE_P(
         testing::Values(false)), // TODO: testing::Bool() after implementing
                                  // communication lowering
     [](const testing::TestParamInfo<std::tuple<InOutMesh, bool>>& info) {
-      const auto& meshes = std::get<0>(info.param);
-      const auto& in_mesh = meshes.first;
-      const auto& out_mesh = meshes.second;
-      const auto enable_hir = std::get<1>(info.param);
-
-      std::stringstream ss;
-      ss << "InMesh";
-      for (auto id : in_mesh.vector()) {
-        ss << "_" << id;
-      }
-      ss << "_OutMesh";
-      for (auto id : out_mesh.vector()) {
-        ss << "_" << id;
-      }
-      ss << (enable_hir ? "_HirEnabled" : "_HirDisabled");
-
-      return ss.str();
+      return nameFromTuple(info.param);
     });

 class LowerSendRecvTest
```

nsarka replied:
That makes sense. In this diff I dont think we need the lambda though, so I updated the PR to just call nameFromTuple in the macro directly.

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -134,18 +180,46 @@ TEST_P(LowerScatterTest, ) {
+        testing::Values(false)), // TODO: testing::Bool() after implementing
+                                 // communication lowering
+    [](const testing::TestParamInfo<std::tuple<InOutMesh, bool>>& info) {
+      const auto& meshes = std::get<0>(info.param);
```
ditto

wujingyue commented on tests/cpp/test_multidevice_lower_communication.cpp:
```diff
@@ -178,17 +252,44 @@ TEST_P(LowerSendRecvTest, ) {
+        testing::Values(false)), // TODO: testing::Bool() after implementing
+                                 // communication lowering
+    [](const testing::TestParamInfo<std::tuple<InOutMesh, bool>>& info) {
+      const auto& meshes = std::get<0>(info.param);
```
ditto

nsarka commented:
!test

nsarka commented:
!test
